client/ping.go

```diff
@@ -71,6 +71,9 @@ func parsePingResponse(cli *Client, resp serverResponse) (types.Ping, error) {
 			ControlAvailable: role == "manager",
 		}
 	}
+	if engineFeatures := resp.header.Get("Engine-Features"); engineFeatures != "" {
+		ping.EngineFeatures = strings.Split(engineFeatures, ",")
```

Comment:
And also, thoughts on the representation in the header?

Reply:
Comma separated looks good to me

Reply:
Taking into account your comment:
> IMO, we should keep the same type.
>Not sure if we'd really want to have a feature like this, but be troublesome to convert when a feature is on by default and it needs an explicit opt-out with a false value. The conversion would then have to know all possible feature flags and append it to the slice if it's not present in daemon's features.

Should we instead do something like `containerd-snapshotter=true,other-feature=false,non-boolean-feature=bork`?

Reply:
Yep

Reply:
>> And also, thoughts on the representation in the header?
>
> Comma separated looks good to me


So the HTTP spec [rfc9110](https://www.rfc-editor.org/rfc/rfc9110.html#name-field-lines-and-combined-fi) allows both "same header multiple times" and "single header, comma-separated", and says that if multiple times, it must be considered the same as a single header, but concatenated (comma-separated).

So;

```
Example-Field: Foo, Bar
Example-Field: Baz
```

Should be considered equivalent to:

```
Example-Field: Foo, Bar, Baz
```

Whitespaces are optional, and should be handled correctly here.

Unfortunately, go's handling of these is a bit dubious; https://go.dev/play/p/y3IYk-6HAZV

```go
package main

import (
	"net/http"
	"os"
)

func main() {
	hdr := http.Header{}

	hdr.Add("Engine-Features", "feature-one")
	hdr.Add("Engine-Features", "feature-two")
	hdr.Add("Engine-Features", "feature-three, feature-four")
	_ = hdr.Write(os.Stdout)

	// Returns all values, but doesn't split commas to separate values :sadpanda:
	values := hdr.Values("Engine-Features")
	t.Log(values)

	// Only returns the first value
	val := hdr.Get("Engine-Features")
	t.Log(val)

	resp := http.Response{
		StatusCode: 200,
		Header:     hdr,
	}
	_ = resp.Write(os.Stdout)
}
```

Reply:
Based on the above, the "easy" way would be to send multiple headers, and let Golang return all the individual ones, but we could consider either, and also split values.

If we _do_ split, we may want to;

- limit number of values  (`strings.SplitN` instead of `strings.Split`) to prevent unbounded splits (although that would only be on some "bad" host)
- trim whitespaces
- perhaps a total limit (to prevent X headers each having X comma-separated values)

We don't expect many values to be present, so we can constraint it to something reasonable.

Reply:
Ah, due to that I ended up implementing the string building/splitting on my own. I think that's a little better for this feature, and we'd probably want to do some normalization/checking anyway so I think this is preferable.

Reply:
Since we're now formatting this such as `Engine-Features: containerd-snapshotter=true,other=false`, we now need features to neither include commas `,` or equals `=` runes. I added some validation to the config features:
```console
# cat /etc/docker/daemon.json
{
        "features": {
                "containerd-snapshotter": true,
                "a,comma": false
        },
        "runtimes": {
                "crun": {
                        "path": "/usr/local/bin/crun"
                }
        }
}
# dockerd -D
unable to configure the Docker daemon with file /etc/docker/daemon.json: merged configuration validation from file and command line flags failed: invalid feature - key cannot contain ',': a,comma
```
and
```console
root@9f4d82f837fb:/go/src/github.com/docker/docker# cat /etc/docker/daemon.json
{
        "features": {
                "containerd-snapshotter": true,
                "an=equals": false
        },
        "runtimes": {
                "crun": {
                        "path": "/usr/local/bin/crun"
                }
        }
}
root@9f4d82f837fb:/go/src/github.com/docker/docker# dockerd -D
unable to configure the Docker daemon with file /etc/docker/daemon.json: merged configuration validation from file and command line flags failed: invalid feature - key cannot contain '=': an=equals
```
I also added checks to the client ping parsing, and on the ping handler (although I guess it could be argued that if we're already validating at the gate, checking when responding to `/_ping` might be redundant.

Reply:
Re:
> - limit number of values (strings.SplitN instead of strings.Split) to prevent unbounded splits (although that would only be on some "bad" host)
> - trim whitespaces

We can definitely do those on the client parsing side, to secure better against bad hosts from blowing up the client's memory.

Reply:
> I  added some validation to the config features:

Good call!

I'm tempted to say that we should follow the same conventions as we use for labels, but I started looking if we actually implemented validation for that. I do recall at the time that we didn't want to make things _too strict_ (other than documenting conventions), but it looks like we have validation on the client side;

```bash
docker create --label 'hello world' alpine
invalid argument "hello world" for "-l, --label" flag: label 'hello world' contains whitespaces
See 'docker create --help'.
```

But not on the API side (sigh);

```bash
curl --unix-socket /var/run/docker.sock -H 'Content-Type: application/json' -X POST -d '{"AttachStderr":true,"AttachStdin":false,"AttachStdout":true,"Cmd":null,"Domainname":"","Entrypoint":null,"Env":null,"HostConfig":{"AutoRemove":false,"Binds":null,"BlkioDeviceReadBps":[],"BlkioDeviceReadIOps":[],"BlkioDeviceWriteBps":[],"BlkioDeviceWriteIOps":[],"BlkioWeight":0,"BlkioWeightDevice":[],"CapAdd":null,"CapDrop":null,"Cgroup":"","CgroupParent":"","CgroupnsMode":"","ConsoleSize":[34,139],"ContainerIDFile":"","CpuCount":0,"CpuPercent":0,"CpuPeriod":0,"CpuQuota":0,"CpuRealtimePeriod":0,"CpuRealtimeRuntime":0,"CpuShares":0,"CpusetCpus":"","CpusetMems":"","DeviceCgroupRules":null,"DeviceRequests":null,"Devices":[],"Dns":[],"DnsOptions":[],"DnsSearch":[],"ExtraHosts":null,"GroupAdd":null,"IOMaximumBandwidth":0,"IOMaximumIOps":0,"IpcMode":"","Isolation":"","Links":null,"LogConfig":{"Config":{},"Type":""},"MaskedPaths":null,"Memory":0,"MemoryReservation":0,"MemorySwap":0,"MemorySwappiness":-1,"NanoCpus":0,"NetworkMode":"default","OomKillDisable":false,"OomScoreAdj":0,"PidMode":"","PidsLimit":0,"PortBindings":{},"Privileged":false,"PublishAllPorts":false,"ReadonlyPaths":null,"ReadonlyRootfs":false,"RestartPolicy":{"MaximumRetryCount":0,"Name":"no"},"SecurityOpt":null,"ShmSize":0,"UTSMode":"","Ulimits":[],"UsernsMode":"","VolumeDriver":"","VolumesFrom":null},"Hostname":"","Image":"alpine","Labels":{"hello world":""},"NetworkingConfig":{"EndpointsConfig":{"default":{"Aliases":null,"DNSNames":null,"DriverOpts":null,"EndpointID":"","Gateway":"","GlobalIPv6Address":"","GlobalIPv6PrefixLen":0,"IPAMConfig":null,"IPAddress":"","IPPrefixLen":0,"IPv6Gateway":"","Links":null,"MacAddress":"","NetworkID":""}}},"OnBuild":null,"OpenStdin":false,"StdinOnce":false,"Tty":false,"User":"","Volumes":{},"WorkingDir":""}' http://localhost/v1.47/containers/create
```